tests/sparse/test_math.py

```diff
@@ -0,0 +1,1274 @@
+
+            self._compile_and_check(
+                [x, y], [psm.dot(x, y)], [x_v, y_v], (Dot, Usmm, UsmmCscDense)
+            )
+
+    def test_sparse_sparse(self):
+        for d1, d2 in [
+            ("float32", "float32"),
+            ("float32", "float64"),
+            ("float64", "float32"),
+            ("float64", "float64"),
+            ("float32", "int16"),
+            ("float32", "complex64"),
+        ]:
+            for x_f, y_f in [
+                ("csc", "csc"),
+                ("csc", "csr"),
+                ("csr", "csc"),
+                ("csr", "csr"),
+            ]:
+                x = SparseTensorType(format=x_f, dtype=d1)("x")
+                y = SparseTensorType(format=x_f, dtype=d2)("x")
+
+                def f_a(x, y):
+                    return x * y
+
+                f_b = pytensor.function([x, y], psm.dot(x, y))
+
+                vx = getattr(self, "x_" + x_f).astype(d1)
+                vy = getattr(self, "y_" + y_f).astype(d2)
+                utt.assert_allclose(f_a(vx, vy).toarray(), f_b(vx, vy))
+
+                f_a = pytensor.function([x, y], psm.dot(x, y).shape)
+
+                def f_b2(x, y):
+                    return (x * y).shape
+
+                assert np.all(f_a(vx, vy) == f_b2(vx, vy))
+                topo = f_a.maker.fgraph.toposort()
+                assert not any(
+                    isinstance(node.op, Dot | Usmm | UsmmCscDense) for node in topo
+                )
+
+    def test_int32_dtype(self):
+        size = 9
+        intX = "int32"
+
+        C = matrix("C", dtype=intX)
+        I = matrix("I", dtype=intX)
+
+        fI = I.flatten()
+        data = pt.ones_like(fI)
+        indptr = pt.arange(data.shape[0] + 1, dtype="int32")
+
+        m1 = CSR(data, fI, indptr, (8, size))
+        m2 = structured_dot(m1, C)
+        y = m2.reshape(shape=(2, 4, 9), ndim=3)
+
+        f = pytensor.function(inputs=[I, C], outputs=y)
+        i = np.asarray([[4, 3, 7, 7], [2, 8, 4, 5]], dtype=intX)
+        a = np.asarray(
+            np.random.default_rng().integers(0, 100, (size, size)), dtype=intX
+        )
+        f(i, a)
+
+
+class TestUsmm:
+    def setup_method(self):
+        x_size = (10, 100)
+        y_size = (100, 200)
+        z_size = (x_size[0], y_size[1])
+
+        self.rng = np.random.default_rng(seed=utt.fetch_seed())
+        self.x = np.asarray(
+            self.rng.binomial(1, 0.5, x_size), dtype=pytensor.config.floatX
+        )
+        self.y = np.asarray(
+            self.rng.uniform(-1, 1, y_size), dtype=pytensor.config.floatX
+        )
+        self.z = np.asarray(
+            self.rng.uniform(-1, 1, z_size), dtype=pytensor.config.floatX
+        )
+
+    @pytest.mark.slow
+    def test_basic(self):
+        def mat(format, name, dtype):
+            if format == "dense":
+                return matrix(name, dtype=dtype)
+            else:
+                return pytensor.sparse.matrix(format, name, dtype=dtype)
+
+        params = product(
```

Comment:
This leads to 2,304 test cases, which seems slightly excessive.

Reply:
only?